src/MeshKernel/CurvilinearGrid.cpp

```diff
@@ -445,8 +462,121 @@ void meshkernel::CurvilinearGrid::ComputeGridNodeTypes()
                 !isBottomLeftFaceValid &&
                 isBottomRightFaceValid)
             {
-                m_gridNodesMask[m][n] = NodeType::UpperLeft;
+                m_gridNodesTypes[m][n] = NodeType::UpperLeft;
             }
         }
     }
 }
+
+void CurvilinearGrid::InsertFace(Point const& point)
+{
+    if (!point.IsValid())
+    {
+        throw std::invalid_argument("CurvilinearGrid::InsertFace: invalid point provided");
+    }
+
+    // Gets the indices of the closest edge to the specified point (they are neighbors by construction)
+    auto const [firstNode, secondNode] = GetEdgeNodeIndices(point);
+
+    if (!firstNode.IsValid() || !secondNode.IsValid())
+    {
+        throw std::invalid_argument("CurvilinearGrid::InsertFace: no valid nodes found");
+    }
+
+    // Compute the grid node types
+    ComputeGridNodeTypes();
+
+    // Add a new edge
+    AddEdge(firstNode, secondNode);
+
+    // Add first node
+    ComputeGridNodeTypes();
+    m_numM = m_gridNodes.size();
+    m_numN = m_gridNodes[0].size();
+    SetFlatCopies();
+}
+
+void CurvilinearGrid::AddEdge(NodeIndices const& firstNode,
+                              NodeIndices const& secondNode)
+{
```

Comment:
this function is a bit long and complex, is it possible to split it into smaller functions?